keps/sig-multicluster/5313-placement-decision-api/README.md

```diff
@@ -0,0 +1,429 @@
+- [ ] (R) Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+Today every multicluster scheduler publishes its own API to convey where a workload should run,
+forcing downstream tools such as GitOps engine, workload orchestrator, progressive rollout controller,
+or AI/ML pipeline having to understand a scheduler specific API.
+
+This KEP introduces a vendor neutral `PlacementDecision` API that standardizes
+the output of multicluster placement calculations.  
+A `PlacementDecision` object is data only: a namespaced list of chosen clusters
+whose `clusterName` values must map one to one to `ClusterProfile` names defined by the
+[ClusterProfile API](https://github.com/kubernetes/enhancements/pull/4322).
+Any scheduler can emit the object and any consumer can watch it.
+Neither side needs to know about the other, enabling true plug and play in the multicluster stack.
+
+![PlacementDecision Flow Diagram](./placementdecision.jpg)
+
+
+## Motivation
+
+A typical multicluster setup:
+
+1. Scheduler: examines the fleet (`ClusterProfile` objects) and decides *where* a workload should land.
+2. Consumer: GitOps engine, workload orchestrator, progressive rollout controller, AI/ML pipeline
+   read that decision and act (usually by creating [Work](https://github.com/kubernetes-sigs/work-api) objects).
+
+Currently every scheduler have its own API for step #1
+so each consumer might need to learn different APIs which slow down integration,
+locking users to specific vendors, and complicate RBAC/validation work.
+
+A standardize unversial `PlacementDecision` API:
+
+* Decouples schedulers from consumers swap either side without rewriting the other.
+* Aligns with the SIG-Multicluster `ClusterProfile` inventory:
+  the decision's cluster keys are guaranteed to exist.
+* Simplifies RBAC: one resource and label schema to secure.
+* Enables GPU aware AI scheduling, cost based placement, and rapid failover.
+
+### Goals
+
+* Define a namespaced scope, minimalistic, data only `PlacementDecision` API that lists selected clusters.
+* Support continuous rescheduling: decision list may be updated.
+* Guarantee that every `clusterName` entry matches a `ClusterProfile.metadata.name` in the same inventory.
+* Guarantee that every `clusterName` entry is in the same namespace as `PlacementDecision.metadata.namespace`.
+* Provide label conventions so consumers can retrieve all slices of one placement.
```

Comment:
I wonder if we actually need this added complexity yet? How many clusters would a placement REALLY need to target? If we get more than a 100, maybe we're not using the right grouping/abstraction?

Reply:
There are fleets running across thousands of clusters, especially in edge and telco cases. While not everyday case, we have customers do push configs to all of those clusters at once.
The core issue with dumping large lists into a single object is that every change results in a large write. K8s API authors already designed APIs like EndpointSlice to handle watch/write churn and workaround etcd limits. So it makes sense to follow established conventions. Expecting users to manually shard their Placement objects to avoid etcd limits or expensive writes feels like a step backward in API design. CC @deads2k